adept/container/actorlearner/learner_container.py

```diff
@@ -0,0 +1,374 @@
+        if args.custom_network:
+            net_cls = REGISTRY.lookup_network(args.custom_network)
+        else:
+            net_cls = ModularNetwork
+        net = net_cls.from_args(
+            args,
+            env_observation_space,
+            output_space,
+            env_gpu_preprocessor,
+            REGISTRY
+        )
+        self.network = net.to(device)
+        # TODO: this is a hack, remove once queuer puts rollouts on the correct device
+        self.network.device = device
+        self.device = device
+        self.network.train()
+
+        # OPTIMIZER
+        def optim_fn(x):
+            return torch.optim.RMSprop(x, lr=args.lr, eps=1e-5, alpha=0.99)
+        if args.nb_learners > 1:
+            self.optimizer = NCCLOptimizer(optim_fn, self.network, self.nb_learners)
+        else:
+            self.optimizer = optim_fn(self.network.parameters())
+
+        # LEARNER / EXP
+        rwd_norm = REGISTRY.lookup_reward_normalizer(
+            args.rwd_norm).from_args(args)
+        actor_cls = REGISTRY.lookup_actor(args.actor_host)
+        builder = actor_cls.exp_spec_builder(
+            env.observation_space,
+            env.action_space,
+            net.internal_space(),
+            args.nb_env * args.nb_learn_batch
+        )
+        w_builder = REGISTRY.lookup_actor(args.actor_worker).exp_spec_builder(
+            env.observation_space,
+            env.action_space,
+            net.internal_space(),
+            args.nb_env
+        )
+        actor = actor_cls.from_args(args, env.action_space)
+        learner = REGISTRY.lookup_learner(args.learner).from_args(args, rwd_norm)
+
+        exp_cls = REGISTRY.lookup_exp(args.exp).from_args(args, builder)
+
+        self.actor = actor
+        self.learner = learner
+        self.exp = exp_cls.from_args(args, builder).to(device)
+
+        # Rank 0 setup, load network/optimizer and create SummaryWriter/Saver
+        if rank == 0:
+            if args.load_network:
+                self.network = self.load_network(self.network, args.load_network)
+                logger.info('Reloaded network from {}'.format(args.load_network))
+            if args.load_optim:
+                self.optimizer = self.load_optim(self.optimizer, args.load_optim)
+                logger.info('Reloaded optimizer from {}'.format(args.load_optim))
+
+            logger.info('Network parameters: ' + str(self.count_parameters(net)))
+            self.summary_writer = SummaryWriter(log_id_dir)
+            self.saver = SimpleModelSaver(log_id_dir)
+
+    def run(self, workers, profile=False):
+        if profile:
+            try:
+                from pyinstrument import Profiler
+            except:
+                raise ImportError('You must install pyinstrument to use profiling.')
+            profiler = Profiler()
+            profiler.start()
+
+        # setup queuer
+        self.rollout_queuer = RolloutQueuerAsync(workers, self.nb_learn_batch, self.rollout_queue_size)
```

Comment:
code smell, can workers be passed to init so that this can go with rest of setup or is there something preventing that?

Reply:
good catch, it doesn't need to be a self variable. but in general workers do need to be created after the learners so learners get gpu access first